src/weather_model_graphs/create/base.py

```diff
@@ -70,27 +75,38 @@ def create_all_graph_components(
     - "containing_rectangle": For each grid node, find the rectangle with 4 mesh nodes as corners
         such that the grid node is contained within it. Connect these 4 (or less along edges)
         mesh nodes to the grid node.
+
+    `projection` should either be a cartopy.crs.CRS or None. This is the projection
+    instance used to transform given lat-lon coords to in-projection Cartesian coordinates.
+    If None the coords are assumed to already be Cartesian.
     """
     graph_components: dict[networkx.DiGraph] = {}
 
-    if len(xy.shape) != 3:
-        raise NotImplementedError(
-            "Mesh coordinates are assumed to lie on a regular grid so that "
-            "the coordinates values are given with an array of shape [2, nx, ny]"
+    assert (
+        len(coords.shape) == 2 and coords.shape[1] == 2
+    ), "Grid node coordinates should be given as an array of shape [num_grid_nodes, 2]."
+
+    if projection is None:
+        logger.debug(
+            "No `projection` given: Assuming `coords` contains in-projection Cartesian coordinates."
+        )
+        xy = coords
+    else:
+        logger.debug(
+            f"`projection` Proj({projection}) given, `coords` treated as lat-lons."
+        )
+        # Convert lat-lon coords to Cartesian xy
+        xyz = projection.transform_points(
+            src_crs=ccrs.PlateCarree(), x=coords[:, 0], y=coords[:, 1]
```

Comment:
Do we need the projections to be cartopy projections or would it suffice to have pyproj projections? Having done a deep dive into cartopy in https://github.com/mllam/mllam-data-prep/issues/33 I think it would be great to keep cartopy out of the loop as long as possible, e.g. only use it for plotting. I think this transformation should also be possible with:
```python
from pyproj import Transformer

transformer = Transformer.from_crs("EPSG:4326", "ESRI:54001")
x, y = transformer.transform(12.4924, 41.8902)
```
The transformer also handles cartopy.crs.CRS if provided, so this approach would just be more general.

Reply:
I think it should be no problem to rely on pyproj over cartopy for this, we just need a way to transform the coordinates. I only want to make sure it is easy for users to specify the projection. Could you give an example of how an input for `projection` at https://github.com/joeloskarsson/weather-model-graphs/blob/4fb69dcf2a91560e5dd80247dd313ec7d88f23c6/src/weather_model_graphs/create/base.py#L40 would look like with pyproj?

I guess it would be a `pyproj.Proj`? Is there a list similar to https://scitools.org.uk/cartopy/docs/v0.15/crs/projections.html for pyproj that we could refer users to for how to construct them?

Reply:
This is a good point @observingClouds, but I agree with @joeloskarsson that allowing people to provide projections as `cartopy.crs.Projection` objects is something people will be familiar with. Could we support both? Or is that going to be too complicated do you think @observingClouds?

If you were to pass in the projection as an argument to `create_all_graph_components(..., crs=...)` what would that look like in this example?

The other thing to consider is that `weather-model-graphs` is built to produce visualisations, but I had made that optional (and therefore put `matplotlib` in the `[visualisation]` optional dependency group in `pyproject.toml` for this reason: https://github.com/mllam/weather-model-graphs/blob/main/pyproject.toml#L23. Maybe if we work out how to support both methods of defining the projection then `cartopy` can go into the `visualisation` dependency group?

Reply:
`Transformer.from_crs()` accepts also cartopy.crs.CRS derived types, so only the function here needs to be changed and the examples can remain the same from what I can see:

```python
from pyproj import Transformer
import cartopy.crs as ccrs
import numpy as np

coords = np.array([[12.4924],[41.8902]])

ccrs.Mercator().transform_points(ccrs.PlateCarree(), x=coords[0], y=coords[1])
# array([[1390647.6067859 , 5116008.22708512,       0.        ]])

Transformer.from_crs(ccrs.PlateCarree(), ccrs.Mercator(), always_xy=True).transform(coords[0], coords[1])
#(array([1390647.6067859]), array([5116008.22708512]))

Transformer.from_crs("EPSG:4326", "EPSG:3857", always_xy=True).transform(coords[0], coords[1])
# (array([1390647.60678589]), array([5144546.10036227])) # slight changes as the standards are differently defined than the general cartopy objects
```

Reply:
@joeloskarsson pyproj does not have a list of available general projections, they are more flexible.

Reply:
> I guess it would be a pyproj.Proj
I have to say I'm not an expert on projections, but if I read https://github.com/joeloskarsson/weather-model-graphs/blob/4fb69dcf2a91560e5dd80247dd313ec7d88f23c6/src/weather_model_graphs/create/base.py#L40 correctly, then a coordinate reference system (CRS) is requested and not a projection as the input is projected to Plate Carree, opposed to just calling a projection that would have src and target CRS already defined.

So maybe `crs` instead of `projection` would be a better keyword argument. But the difference is small, particularly because also _projected coordinate reference systems_ exist. In the current form of the function call, the input would be of type pyproj.crs.CRS or cartopy.crs.CRS (as it is already).

Does this make sense @joeloskarsson ?

Reply:
> So maybe `crs` instead of `projection` would be a better keyword argument. But the difference is small, particularly because also _projected coordinate reference systems_ exist

I think this makes sense. I was just reading up on what exactly at CRS is to make sure I am using the nomenclature correctly (https://docs.qgis.org/3.34/en/docs/gentle_gis_introduction/coordinate_reference_systems.html#coordinate-reference-system-crs-in-detail). I think it would be clearer to say that the argument defines the coordinate reference system, whether that is not projected (i.e. describes positions on the Earth for example using lat/lons and specific Earth properties) or projected (i.e. builds on a not projected coordinate reference system by applying a specific projection)

Reply:
Thanks for explaining and good suggestions! 

> then a coordinate reference system (CRS) is requested and not a projection as the input is projected to Plate Carree, opposed to just calling a projection that would have src and target CRS already defined.

This is a very good point. I think I have been very sloppy with the differences between crs and projection both when talking about this and thinking about it :sweat_smile: It is absolutely true that the argument now is a crs and, when given, the coords input is always transformed from PlateCarree to this crs.  

I think it is important to allow for cartopy.crs.CRS as options, as that is a lot easier for most users. But as pyproj Transformer.from_crs() accepts also that I think it would be great to change to using that. Then the pyproj only gives additional flexibility which is great!

As for the keyword name I agree that crs would be more accurate then projection. It seems however a bit confusing to me that the logic would be 1. If crs == None: assume that coordinates are already given in the desired coordinate system for graph creation 2. if crs != None: Assume coordinates are lat-lons and transform them to the given crs.
I propose to instead have two keyword arguments: coords_crs and graph_crs, where they both default to None (or maybe coords_src defaults to Plate Carree?). When both are given the coords are transformed from coords_crs to the graph_crs, where the graph creation is carried out. That also adds even more flexibility, as it allows for transforming from arbitrary crs:s, not just lat-lons.

Reply:
> I propose to instead have two keyword arguments: coords_crs and graph_crs

I like this suggestion as it is more explicit and adds flexibility.

Reply:
Thanks @joeloskarsson for the changes! LGTM!

Reply:
Looks good to me too @joeloskarsson. I'll resolve this conversation :)